File: frontend/packages/contracts/src/event/event.dto.ts
import { z } from "zod";

import { deviceDto, type TraitValue } from "../device";

export const deviceStateEvent = z.object({
  type: z.literal("device.state"),
  device: deviceDto,
});

export const deviceListEvent = z.object({
  type: z.literal("device.list"),
  hub_id: z.uuid(),
  devices: z.array(deviceDto),
});

export const hubPresenceEvent = z.object({
  type: z.literal("hub.presence"),
  hub_id: z.uuid(),
  online: z.boolean(),
});

export const commandRejectedEvent = z.object({
  type: z.literal("command.rejected"),
  device_id: z.uuid(),
  reason: z.string(),
  message: z.string(),
});

export const avaEvent = z.discriminatedUnion("type", [
  deviceStateEvent,
  deviceListEvent,
  hubPresenceEvent,
  commandRejectedEvent,
]);

export type DeviceCommandFrame = {
  type: "device.command";
  device_id: string;
  trait: string;
  value: TraitValue;
};

export type CommandRejectedEvent = z.infer<typeof commandRejectedEvent>;
export type DeviceStateEvent = z.infer<typeof deviceStateEvent>;
export type DeviceListEvent = z.infer<typeof deviceListEvent>;
export type HubPresenceEvent = z.infer<typeof hubPresenceEvent>;
export type AvaEvent = z.infer<typeof avaEvent>;

File: backend/services/api/internal/dto/event_dto.go
package dto

import (
	"ava/pkg/wire"

	"github.com/google/uuid"
)

const (
	EventDeviceState = "device.state"
	EventDeviceList  = "device.list"
	EventHubPresence = "hub.presence"

	EventDeviceCommand   = "device.command"
	EventCommandRejected = "command.rejected"
)

type DeviceCommandFrame struct {
	Type     string     `json:"type"`
	DeviceID uuid.UUID  `json:"device_id"`
	Trait    wire.Trait `json:"trait"`
	Value    wire.Value `json:"value"`
}

type CommandRejectedEvent struct {
	Type     string    `json:"type"`
	DeviceID uuid.UUID `json:"device_id"`
	Reason   string    `json:"reason"`
	Message  string    `json:"message"`
}

func NewCommandRejectedEvent(deviceID uuid.UUID, reason, message string) CommandRejectedEvent {
	return CommandRejectedEvent{
		Type:     EventCommandRejected,
		DeviceID: deviceID,
		Reason:   reason,
		Message:  message,
	}
}

type DeviceStateEvent struct {
	Type   string          `json:"type"`
	Device *DeviceResponse `json:"device"`
}

func NewDeviceStateEvent(device *DeviceResponse) DeviceStateEvent {
	return DeviceStateEvent{Type: EventDeviceState, Device: device}
}

type DeviceListEvent struct {
	Type    string            `json:"type"`
	HubID   uuid.UUID         `json:"hub_id"`
	Devices []*DeviceResponse `json:"devices"`
}

func NewDeviceListEvent(hubID uuid.UUID, devices []*DeviceResponse) DeviceListEvent {
	if devices == nil {
		devices = []*DeviceResponse{}
	}

	return DeviceListEvent{Type: EventDeviceList, HubID: hubID, Devices: devices}
}

type HubPresenceEvent struct {
	Type   string    `json:"type"`
	HubID  uuid.UUID `json:"hub_id"`
	Online bool      `json:"online"`
}

func NewHubPresenceEvent(hubID uuid.UUID, online bool) HubPresenceEvent {
	return HubPresenceEvent{Type: EventHubPresence, HubID: hubID, Online: online}
}
